### PR TITLE
Use `sr-tool` for Release

### PR DESCRIPTION
**Pull request by fewensa**
- Automatic test and create release
- Compile wasm using srtool
- A simple guide written in Chinese

ref:
https://github.com/darwinia-network/darwinia/issues/627

### CODE REVIEW

File: .maintain/README_zh_CN.md
Comment (HackFisher) [inline .maintain/README_zh_CN.md:96]: 1. 目前里程碑模式(release)下Github Action会自动输出srtool编译的wasm binary吗? 如果有的话，输出文件在那里。
2. 如果不会自动输出的话，提交binary的人是不是需要根据paritytech/srtool 来配置环境，进行手动的编译?
3. 验证的人是不是也是根据paritytech/srtool 来配置环境，进行手动的编译，然后验证Hash?

File: .maintain/README_zh_CN.md
Comment (HackFisher) [inline .maintain/README_zh_CN.md:108]: 如果我现在尝试的话，是不是应该用chevdor/srtool?

File: .maintain/README_zh_CN.md
Comment (fewensa) [reply .maintain/README_zh_CN.md:96]: 1. 会自动输出 wasm binary 文件, 并且发布到 Github release 页面, 其效果类似于 [v0.9.5](https://github.com/darwinia-network/darwinia/releases/tag/v0.9.5) Assets. 在发布后, 会自动上传挂载 wasm binary
2. 会自动输出, 无需手动编译提交
3. 验证的人有两种方式进行验证
    1. 在目前策略中, 有规划将 hash 自动输出到 release 页面, 验证人可直接从 release 页面中查找 (目前尚未实现, 存在技术问题, 将会和 Way 进行沟通如何获取)
    2. 验证人也可以透过 paritytech/srtool 来进行手动编译, 这是文档中尚未进行阐述的

File: .maintain/README_zh_CN.md
Comment (fewensa) [reply .maintain/README_zh_CN.md:108]: 实际上, 采用 paritytech/srtool 或者 chevdor/srtool 没有本质上的区别, 二者的目的均相同, 使用相同的系统, 固定的 rustc 版本对 runtime 进行编译. chevdor/srtool 相对于 paritytech/srtool 而言, 只是更新频率相对较快 (更新频率只针对 rustc 版本). 而 paritytech/srtool 只有一个 [nightly-2020-10-27](https://hub.docker.com/r/paritytech/srtool/tags?page=1&ordering=last_updated) 版本, 二者内部实现方式均相同, 使用 `cargo build` 进行编译.

因此就目前情况而言, 建议使用  paritytech/srtool 这样不管是自己编译还是 ci 编译 都使用的是统一的版本, 如果是使用  chevdor/srtool 不同的人选择了不同的 rustc 版本, 反而有可能会导致 hash 的不同.

File: .maintain/README_zh_CN.md
Comment (fewensa) [reply .maintain/README_zh_CN.md:96]: 验证的人, 验证方式 1 已实现, 将会自动捕获 srtool 所生成的 Proposal hash 将其挂载到 Assets
文件名格式:

PROPOSAL_HASH.proposal.darwinia.txt
PROPOSAL_HASH.proposal.crab.txt

PROPOSAL_HASH 便是 hash 值, 文件内容也是 hash 值

**Review by HackFisher — APPROVED** [review]

**Comment by AurevoirXavier** [comment]
Is this working?

Or only the tag will trigger this.

**Comment by fewensa** [comment]
> 
> 
> Is this working?
> 
> Or only the tag will trigger this.

Yes, it's working. I have already tested.
when push code to master branch trigger `cargo test`
when create a new tag will build darwinia binary push to docker registry and build darwinia/crab wasm binary. all binary mount to github release page.

before this pr merge, I'll do a last test in my repository. and reference this link.

**Comment by AurevoirXavier** [comment]
> > Is this working?
> > Or only the tag will trigger this.
> 
> Yes, it's working. I have already tested.
> when push code to master branch trigger `cargo test`
> when create a new tag will build darwinia binary push to docker registry and build darwinia/crab wasm binary. all binary mount to github release page.
> 
> before this pr merge, I'll do a last test in my repository. and reference this link.

Yes. I want to see the tag process.

**Comment by fewensa** [comment]
pass the final test, see this link https://github.com/fewensa/darwinia/releases/tag/v0.9.5-7

File: .github/workflows/ci.yml
Comment (AurevoirXavier) [inline .github/workflows/ci.yml:37]: Is this line will affect the sr-tool?

Comment (fewensa) [reply .github/workflows/ci.yml:37]: no, srtool only build release. is there
https://github.com/fewensa/darwinia/blob/features%2Fci/.github/workflows/release.yml#L36

Comment (AurevoirXavier) [reply .github/workflows/ci.yml:37]: So nightly is fine.

In #636

Comment (fewensa) [reply .github/workflows/ci.yml:37]: yeah, in .github/workflows/ci.yml use nightly is fine, but when you create a new tag, will run .github/workflows/release.yml

you can see that
https://github.com/darwinia-network/darwinia/runs/2146405787?check_suite_focus=true#step:5:37
currently release build use `nightly-2020-10-06` so when you push a new tag, this version may not work.


your change not effect commit to push master, but your code may not support `nightly-2020-10-06` (currently code), and `nightly-2020-10-27` (this pr, srtool use this version). 


in this pr, .github/workflows/release.yml build docker images can use `nightly` version, but paritytech/srtool only support `nightly-2020-10-27`

Comment (AurevoirXavier) [reply .github/workflows/ci.yml:37]: Will check.

**Comment by HackFisher** [comment]
At least 2 approving reviews are required before merging these PRs, please proceed for the approve or merge if there are no more changes required.

The same for other pull requests.

**Comment by AurevoirXavier** [comment]
> At least 2 approving reviews are required before merging these PRs, please proceed for the approve or merge if there are no more changes required.
> 
> The same for other pull requests.

Review-ing...

**Comment by fewensa** [comment]
I'm test currently, let's me do some change.

**Review by AurevoirXavier — CHANGES_REQUESTED** [review]
https://github.com/fewensa/darwinia/releases/download/v0.9.5-7/darwinia_x86_64-linux-gnu

Is this binary support universal linux or just ubuntu-latest?

Also the file size is too large. Please tar with bz2.

**Comment by HackFisher** [comment]
Related https://github.com/darwinia-network/darwinia/issues/608#issuecomment-770589007

File: .maintain/docker/Dockerfile.aarch64-linux-gnu
Comment (AurevoirXavier) [inline .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: Please don't remove this line. 

Will slow down the speed in local.

Just comment it.

Comment (AurevoirXavier) [reply .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: Okay.

Another file named chn means CHINA?

**Comment by fewensa** [comment]
All question solved by the new commit. and have some new features,

1. the release.yml enable multi step to build release
2. change paritytech/srtool to chevdor/srtool
3. all toolchain use nightly-2021-03-15

@AurevoirXavier
1. the new commit build binary is support universal linux
2. the upload binary use bz2 package
3. update rust toolchain to nightly-2021-03-15, so support pr #636 

@HackFisher 
About related url, I think @AurevoirXavier  says is right, so now also centos:7

File: .maintain/docker/Dockerfile.aarch64-linux-gnu
Comment (fewensa) [reply .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: I move local file to Dockerfile.mirror-chn.aarch64-linux-gnu, because in github action my not need change mirro.

Comment (fewensa) [reply .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: Yes, it's China

File: .maintain/setup-darwinia.sh
Comment (fewensa) [inline .maintain/setup-darwinia.sh:61]: There I have a question for @AurevoirXavier , in .maintain/utility/release script, after build use aarch64 why copy file use x86_64 folder? 
https://github.com/darwinia-network/darwinia/blob/d1247dc8697bbba397d96b29e40e95106b29cd18/.maintain/utility/release#L37-L39
It's a bug? if that, I can release this comment to support aarch64 build in this script

Comment (AurevoirXavier) [reply .maintain/setup-darwinia.sh:61]: Seems we made a mistake here.

Comment (AurevoirXavier) [reply .maintain/setup-darwinia.sh:61]: And could you remove the old releases' aarch64 binary.

Comment (fewensa) [reply .maintain/setup-darwinia.sh:61]: ok, I'll do last test in my repository enable aarch64, and update this script, when build finish, I'll post release url to here.

File: .maintain/docker/Dockerfile.aarch64-linux-gnu
Comment (wi1dcard) [reply .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: It is not allowed using any kind of mirror on CI/CD, since we're not able to determine if the mirror can be trusted. Even if it is maintained by any well-known organization, the mirror still got a chance having some technical issues such as lagging behind which could potentially cause discrepancy of build results and artifacts.

File: .maintain/setup-darwinia.sh
Comment (fewensa) [reply .maintain/setup-darwinia.sh:61]: > 
> 
> And could you remove the old releases' aarch64 binary? If you don't have the permission let me do that.

You can do this, I don't know how to remove that

File: .maintain/docker/Dockerfile.aarch64-linux-gnu
Comment (fewensa) [reply .maintain/docker/Dockerfile.aarch64-linux-gnu:24]: yeah, this pr i removed all used Dockerfile mirror, like this reviews code. and back old file to Dockerfile.mirror-chn.xxx

**Comment by fewensa** [comment]
complete test, see that https://github.com/fewensa/darwinia/releases/tag/v0.9.5-9

**Comment by AurevoirXavier** [comment]
Switch toolchain to nightly-2021-03-01

**Comment by fewensa** [comment]
> nightly-2021-03-01

No, can not change it. the chevdor/srtool don't have this version

https://hub.docker.com/r/chevdor/srtool/tags?page=1&ordering=last_updated&name=nightly-2021-03

It's have nightly-2021-03-15

**Comment by AurevoirXavier** [comment]
But the CI failed.

Due to https://github.com/rust-lang/rust/issues/83291

**Comment by AurevoirXavier** [comment]
Seriously?

CI takes 3 hours.

@wi1dcard seems something wrong with the slack API?

**Comment by AurevoirXavier** [comment]
Still failed.

**Comment by fewensa** [comment]
> 
> 
> Still failed.

you can see that,
https://github.com/darwinia-network/darwinia/actions/runs/737408184

the compile is success, failed is build image, the reason is can not get `secrets`, may in pull request no permission to get this.

**Comment by fewensa** [comment]
Build success.

ci.yml
https://github.com/darwinia-network/darwinia/actions/runs/737700160
trigger by push to to master or pull request to master

release.yml
https://github.com/darwinia-network/darwinia/actions/runs/737408184
trigger by create new tag (must add `v` prefix to tag name). and need update CHANGELOG.md follow [keep a changelog](https://keepachangelog.com/en/1.0.0/) style
> this test failed, because fork repository create pull request can not get secrets, but compile is success.

and currently code have some comment, that code will be use in future, the new srtool version have some change.

**Review by AurevoirXavier — APPROVED** [review]

**Pull request merged**
